intel-sgx/insecure-time/src/lib.rs

```diff
@@ -0,0 +1,887 @@
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.eq(&o)
+    }
+}
+
+impl PartialOrd<Ticks> for Ticks {
+    fn partial_cmp(&self, other: &Ticks) -> Option<cmp::Ordering> {
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.partial_cmp(&o)
+    }
+}
+
+impl Add for Ticks {
+    type Output = Result<Ticks, Error>;
+
+    fn add(self, other: Self) -> Self::Output {
+        let t0 = self.0.load(Ordering::Relaxed);
+        let t1 = other.0.load(Ordering::Relaxed);
+        t0.checked_add(t1)
+            .ok_or(Error::TypeOverflow)
+            .map(|sum| Ticks(AtomicU64::new(sum)))
+    }
+}
+
+impl Display for Ticks {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{} ticks", self.0.load(Ordering::Relaxed))
+    }
+}
+
+impl Ticks {
+    pub fn new(t: u64) -> Self {
+        Ticks(AtomicU64::new(t))
+    }
+
+    pub const fn max() -> Self {
+        Ticks(AtomicU64::new(u64::MAX))
+    }
+
+    pub fn now() -> Self {
+        // The RDTSC instruction reads the time-stamp counter and is guaranteed to
+        // return a monotonically increasing unique value whenever executed, except
+        // for a 64-bit counter wraparound. Intel guarantees that the time-stamp
+        // counter will not wraparound within 10 years after being reset. The period
+        // for counter wrap is longer for Pentium 4, Intel Xeon, P6 family, and
+        // Pentium processors.
+        // Source: Intel x86 manual Volume 3 Chapter 19.17 (Time-stamp counter)
+        //
+        // However, note that an attacker may arbitarily set this value on the
+        // host/hypervisor
+        Ticks(Rdtscp::read().into())
+    }
+
+    pub fn abs_diff(&self, t1: &Ticks) -> Ticks {
+        let t0 = self.0.load(Ordering::Relaxed);
+        let t1 = t1.0.load(Ordering::Relaxed);
+
+        if t1 > t0 {
+            Ticks::new(t1 - t0)
+        } else {
+            Ticks::new(t0 - t1)
+        }
+    }
+
+    fn set(&self, t: Ticks) {
+        let t = t.0.load(Ordering::Relaxed);
+        self.0.store(t, Ordering::Relaxed);
+    }
+
+    pub fn from_duration(duration: Duration, freq: &Freq) -> Result<Self, Error> {
+        let freq = freq.as_u64();
+        let ticks_secs = duration.as_secs().checked_mul(freq).ok_or(Error::TypeOverflow)?;
+        let ticks_nsecs = (duration.subsec_nanos() as u64)
+            .checked_mul(freq).ok_or(Error::TypeOverflow)? / NANOS_PER_SEC;
+        Ok(Ticks::new(ticks_secs + ticks_nsecs))
+    }
+
+    pub fn as_duration_ex(&self, freq: &Freq) -> Duration {
+        let freq = freq.as_u64();
+        let ticks = self.0.load(Ordering::Relaxed);
+
+        let time_secs = ticks / freq;
+        let time_nsecs = (ticks % freq * NANOS_PER_SEC) / freq;
+        let time_nsecs: u32 = time_nsecs.try_into().expect("must be smaller than 1sec");
+
+        Duration::new(time_secs, time_nsecs)
```

Comment:
Actually, you could only compute `time_nsecs` and use [`Duration::from_nanos`](https://doc.rust-lang.org/std/time/struct.Duration.html#method.from_nanos)

Reply:
I assume you're thinking of:
```
pub fn as_duration_ex(&self, freq: &Freq) -> Duration {
    let freq = freq.as_u64();
    let ticks = self.0.load(Ordering::Relaxed);
    Duration::from_nanos(ticks * NANOS_PER_SEC / freq)
}
```
I'd rather not do that. Doing `ticks * NANOS_PER_SEC` could result in an overflow.